third_party/move/move-vm/runtime/src/runtime_ref_checks.rs

```diff
@@ -0,0 +1,1828 @@
+            | ImmBorrowLoc(_)
+            | ImmBorrowField(_)
+            | MutBorrowField(_)
+            | ImmBorrowFieldGeneric(_)
+            | MutBorrowFieldGeneric(_)
+            | PackClosure(..)
+            | PackClosureGeneric(..)
+            | Pack(_)
+            | PackGeneric(_)
+            | Unpack(_)
+            | UnpackGeneric(_)
+            | WriteRef
+            | CastU8
+            | CastU16
+            | CastU32
+            | CastU64
+            | CastU128
+            | CastU256
+            | Add
+            | Sub
+            | Mul
+            | Mod
+            | Div
+            | BitOr
+            | BitAnd
+            | Xor
+            | Or
+            | And
+            | Shl
+            | Shr
+            | Lt
+            | Le
+            | Gt
+            | Ge
+            | Eq
+            | Neq
+            | MutBorrowGlobal(_)
+            | ImmBorrowGlobal(_)
+            | MutBorrowGlobalGeneric(_)
+            | ImmBorrowGlobalGeneric(_)
+            | Exists(_)
+            | ExistsGeneric(_)
+            | MoveTo(_)
+            | MoveToGeneric(_)
+            | MoveFrom(_)
+            | MoveFromGeneric(_)
+            | FreezeRef
+            | Nop
+            | Not
+            | VecPack(_, _)
+            | VecLen(_)
+            | VecImmBorrow(_)
+            | VecMutBorrow(_)
+            | VecPushBack(_)
+            | VecPopBack(_)
+            | VecUnpack(_, _)
+            | VecSwap(_)
+            | PackVariant(_)
+            | PackVariantGeneric(_)
+            | UnpackVariant(_)
+            | UnpackVariantGeneric(_)
+            | TestVariant(_)
+            | TestVariantGeneric(_)
+            | MutBorrowVariantField(_)
+            | MutBorrowVariantFieldGeneric(_)
+            | ImmBorrowVariantField(_)
+            | ImmBorrowVariantFieldGeneric(_) => {
+                // handled in `post_execution_transition`
+            },
+        };
+        Ok(())
+    }
+
+    fn post_execution_transition(
+        frame: &Frame,
+        instruction: &Bytecode,
+        ref_state: &mut RefCheckState,
+        ty_cache: &mut FrameTypeCache,
+    ) -> PartialVMResult<()> {
+        use Bytecode::*;
+        match instruction {
+            Pop => {
+                let top = ref_state.pop_from_shadow_stack()?;
+                if let Value::Ref(ref_id) = top {
+                    ref_state.purge_reference(ref_id)?;
+                }
+            },
+            Ret | BrTrue(_) | BrFalse(_) | Branch(_) => {
+                // not reachable here, transition is handled in `pre_execution_transition`
+            },
+            CastU8 | CastU16 | CastU32 | CastU64 | CastU128 | CastU256 | Not | Nop | Exists(_)
+            | ExistsGeneric(_) => {
+                // no-op
+            },
+            LdU8(_) | LdU16(_) | LdU32(_) | LdU64(_) | LdU128(_) | LdU256(_) | LdConst(_)
+            | LdTrue | LdFalse => {
+                ref_state.push_non_refs_to_shadow_stack(1);
```

Comment:
Still going through the code, might not see all the nuance, but does it make sense to push non refs at all? We run paranoid mode which check stack balance and types, so running reference safety after will only see valid stack transitions -> we can just use refs only?

Reply:
I did consider this optimization, but I remember there were some interesting corner cases: `Eq` and `Neq` take the top two values on the stack, which can either be both references or both non-references. If there are two references on top of the reference stack (in the world where non refs are not pushed), should we pop them or not? 

But I think we can certainly do optimizations around this idea: collate all the consecutive non-refs on the stack into one "counter" value, use existing type stack from paranoid mode which breaks some abstractions and makes ref checker explicitly dependent on runtime type checker).

I would defer all these to a follow up optimization-focused PR.